Enhancement/mobile drawer

HendrikPN opened the pull request:
This PR closes #27 
I added a [`SwipeableDrawer`](https://material-ui.com/api/swipeable-drawer/) to the `TopicDrawer` and `Settings`.

As you already mentioned, this will also close #22 and #23 since they are contained here.

Note that the permanent drawers (on larger screens) are still only a `Paper` which do not remain in the view. Should we change that?

HendrikPN committed:
Added overlay text to Hero

+ removed `About` text from Drawer
+ added link to Header

HendrikPN committed:
Added Features and Goals to landing page

Additional mods:
+ improved overlay from previous commit
+ updated get_started

HendrikPN committed:
Added small features for landing page

+ image credits on hero
+ divider between feature cards on mobile version
+ Features&Goals section on landing page vanishes if search is applied

HendrikPN committed:
Added github and twitter link to landing

Additional adjustments:
+ added twitter icon
+ update Typography variants
+ removed some errors

HendrikPN committed:
Merge remote-tracking branch 'origin/master' into enhancement/landing_page

HendrikPN committed:
Added prototype images for landing page

credits @LeaMarion

HendrikPN committed:
Added Environment Detail View

+ updated search (both for categories & environments)
+ removed README from `EnvironmentItem`
+ removed some unused imports/styles and cleaned up the repo.

HendrikPN committed:
Added mobile topic drawer

Topic drawer is now swipeable on mobile devices.

HendrikPN committed:
adapted search for drawer

search is now included to the mobile drawer on `xs` screen sizes.

HendrikPN committed:
Added favicon

HendrikPN committed:
added mobile drawer for settings

Settings drawer is now swipeable on mobile devices.

erik-sn commented on src/components/environment/EnvironmentDetail.js:
```diff
@@ -0,0 +1,207 @@
+        filePath = currentAvatar.filePath.replace(constants.UPLOAD_URL, '');
+      }
+    }
+    var converter = new showdown.Converter();
```
As is this will make a new converter every time anything changes in this render which seems like it could be a lot & often. I think that you can probably do this in the class' constructor rather than the render.

HendrikPN replied:
done

erik-sn commented on src/components/header/SearchBar.js:
```diff
@@ -49,48 +70,145 @@ const styles = theme => ({
   }
 
+  handleExpandMore = () => {
+    var showMore = this.state.shownEnv;
```
If you need to mutate a variable use `let` over `var`.

However this is also mutating state which we don't want to do. Try something like:
```
this.setState({ shownEnv: this.state.shownEnv + modDisplay });
```

HendrikPN replied:
done

erik-sn commented on src/components/header/SearchBar.js:
```diff
@@ -49,48 +70,145 @@ const styles = theme => ({
+  };
+
+  handleExpandLess = () => {
+    var showLess = this.state.shownEnv;
```
Same here

HendrikPN replied:
done

erik-sn commented on src/components/header/SearchBar.js:
```diff
@@ -49,48 +70,145 @@ const styles = theme => ({
-    const { searchValue } = this.state;
+    const { searchValue, open } = this.state;
     const { classes } = this.props;
+    var environments = this.props.searchedEnvironments
```
`let` instead off `var`

HendrikPN replied:
done

erik-sn commented on src/components/header/SearchBar.js:
```diff
@@ -49,48 +70,145 @@ const styles = theme => ({
+      : this.props.environments;
+    const all = true ? this.state.shownEnv >= environments.length : false;
+    const none = true ? this.state.shownEnv <= modDisplay : false;
+    environments = environments.slice(0, this.state.shownEnv);
```
possibly give `environments` to `visibleEnvironments` so you can use `const` instead of `var`. Also it's a bit more descriptive.

HendrikPN replied:
done

erik-sn commented on src/components/header/SearchBar.js:
```diff
@@ -49,48 +70,145 @@ const styles = theme => ({
     this.state = {
       searchValue: '',
+      open: false,
+      shownEnv: modDisplay,
```
consider changing this to something like `visibleEnvironmentCount`. More descriptive & explicit.

HendrikPN replied:
done

erik-sn commented on src/components/home/DrawerContent.js:
```diff
@@ -0,0 +1,140 @@
+    super(props);
+    const parentTopics = this.props.topics.filter(topic => !topic.parentTopic);
+    this.state = {
+      open: Array.apply(null, Array(parentTopics.length)).map(() => false),
```
`open` is vague, try to give this a more descriptive name.

HendrikPN replied:
done

erik-sn commented on src/components/home/DrawerContent.js:
```diff
@@ -0,0 +1,140 @@
+  resetCategorizedEnvironments: PropTypes.func.isRequired,
+};
+
+// const mapStateToProps = state => ({
```
can remove this I think

HendrikPN replied:
done

erik-sn commented on src/components/home/EnvironmentItem.jsx:
```diff
@@ -122,7 +82,7 @@ class EnvironmentItem extends Component {
+    const { owner, htmlUrl, gym, fork } = this.props.environment.repository;
     const { name, description, scigym, tags, topic, currentAvatar } = this.props.environment;
     const { classes } = this.props;
     var filePath = constants.SCIGYM_LOGO;
```
I'd scrape through everything looking to remove `var`s anywhere they are.

erik-sn reviewed (changes requested):
Generally it looks good! I've noticed your `render` methods are pretty long so look for opportunities to break these components into smaller chunks. Material-UI does tend to result in these types of things.

HendrikPN committed:
Restructuring components

+ Code has been restructured for readability. Long render methods are avoided mostly.
+ Added subdirectories where useful

HendrikPN committed:
Corrections according to review of #28

+ replace `var` by `let`
+ removed `converter` from `render`
+ improved `showMore` and `showLess` code
+ improved code readability

HendrikPN commented:
Thanks for the review. I have restructured the code quite a bit and taken heed of your comments. The current code readability should have been largely improved.

HendrikPN commented:
As you already mentioned, this will also close PR #22 and #23 since they are contained here.
